File: backend/app/schemas/pgr_governance.py
from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class PGRDocumentApprovalCreate(BaseModel):
    cnpj_id: UUID
    org_unit_id: UUID | None = None
    document_scope: str = Field(default="inventory")
    version_label: str | None = None
    statement: str | None = None
    notes: str | None = None
    effective_from: datetime | None = None
    review_due_at: datetime | None = None


class PGRDocumentApprovalOut(BaseModel):
    id: UUID
    tenant_id: UUID
    cnpj_id: UUID
    org_unit_id: UUID | None = None
    document_scope: str
    version_label: str
    status: str
    statement: str
    notes: str | None = None
    approver_name: str
    approver_role: str | None = None
    approver_email: str | None = None
    effective_from: datetime
    review_due_at: datetime | None = None
    approved_at: datetime
    inventory_item_count: int
    snapshot_hash: str
    snapshot_json: dict[str, Any]
    created_at: datetime
    updated_at: datetime


class ErgonomicAssessmentCreate(BaseModel):
    cnpj_id: UUID
    org_unit_id: UUID | None = None
    assessment_type: str = Field(default="AEP")
    title: str
    process_name: str | None = None
    activity_name: str | None = None
    position_name: str | None = None
    workstation_name: str | None = None
    demand_summary: str | None = None
    conditions_summary: str | None = None
    psychosocial_factors: list[str] = []
    findings: list[str] = []
    recommendations: list[str] = []
    traceability: dict[str, Any] = {}
    review_due_at: datetime | None = None


class ErgonomicAssessmentUpdate(BaseModel):
    assessment_type: str | None = None
    title: str | None = None
    status: str | None = None
    process_name: str | None = None
    activity_name: str | None = None
    position_name: str | None = None
    workstation_name: str | None = None
    demand_summary: str | None = None
    conditions_summary: str | None = None
    psychosocial_factors: list[str] | None = None
    findings: list[str] | None = None
    recommendations: list[str] | None = None
    traceability: dict[str, Any] | None = None
    review_due_at: datetime | None = None
    approval_notes: str | None = None


class ErgonomicAssessmentApprove(BaseModel):
    approval_notes: str | None = None


class ErgonomicAssessmentOut(BaseModel):
    id: UUID
    tenant_id: UUID
    cnpj_id: UUID
    org_unit_id: UUID | None = None
    assessment_type: str
    title: str
    status: str
    process_name: str | None = None
    activity_name: str | None = None
    position_name: str | None = None
    workstation_name: str | None = None
    demand_summary: str | None = None
    conditions_summary: str | None = None
    psychosocial_factors: list[str] = []
    findings: list[str] = []
    recommendations: list[str] = []
    traceability: dict[str, Any] = {}
    reviewed_at: datetime | None = None
    review_due_at: datetime | None = None
    approved_at: datetime | None = None
    approval_notes: str | None = None
    created_at: datetime
    updated_at: datetime

File: frontend/src/lib/api/pgr-governance.ts
import { apiFetch } from "@/lib/api/client";

export type PGRDocumentApprovalOut = {
  id: string;
  tenant_id: string;
  cnpj_id: string;
  org_unit_id?: string | null;
  document_scope: string;
  version_label: string;
  status: string;
  statement: string;
  notes?: string | null;
  approver_name: string;
  approver_role?: string | null;
  approver_email?: string | null;
  effective_from: string;
  review_due_at?: string | null;
  approved_at: string;
  inventory_item_count: number;
  snapshot_hash: string;
  snapshot_json: Record<string, any>;
  created_at: string;
  updated_at: string;
};

export type ErgonomicAssessmentOut = {
  id: string;
  tenant_id: string;
  cnpj_id: string;
  org_unit_id?: string | null;
  assessment_type: "AEP" | "AET" | string;
  title: string;
  status: string;
  process_name?: string | null;
  activity_name?: string | null;
  position_name?: string | null;
  workstation_name?: string | null;
  demand_summary?: string | null;
  conditions_summary?: string | null;
  psychosocial_factors: string[];
  findings: string[];
  recommendations: string[];
  traceability: Record<string, any>;
  reviewed_at?: string | null;
  review_due_at?: string | null;
  approved_at?: string | null;
  approval_notes?: string | null;
  created_at: string;
  updated_at: string;
};

export async function listPgrApprovals(params?: { cnpj_id?: string; org_unit_id?: string; status?: string; limit?: number; offset?: number }) {
  const qs = new URLSearchParams();
  if (params?.cnpj_id) qs.set("cnpj_id", params.cnpj_id);
  if (params?.org_unit_id) qs.set("org_unit_id", params.org_unit_id);
  if (params?.status) qs.set("status", params.status);
  qs.set("limit", String(params?.limit ?? 50));
  qs.set("offset", String(params?.offset ?? 0));
  return apiFetch<{ items: PGRDocumentApprovalOut[]; total: number }>("console", `/pgr/approvals?${qs.toString()}`);
}

export async function createPgrApproval(payload: Record<string, any>) {
  return apiFetch<PGRDocumentApprovalOut>("console", "/pgr/approvals", { method: "POST", body: JSON.stringify(payload) });
}

export async function listErgonomics(params?: { cnpj_id?: string; org_unit_id?: string; assessment_type?: string; status?: string; limit?: number; offset?: number }) {
  const qs = new URLSearchParams();
  if (params?.cnpj_id) qs.set("cnpj_id", params.cnpj_id);
  if (params?.org_unit_id) qs.set("org_unit_id", params.org_unit_id);
  if (params?.assessment_type) qs.set("assessment_type", params.assessment_type);
  if (params?.status) qs.set("status", params.status);
  qs.set("limit", String(params?.limit ?? 100));
  qs.set("offset", String(params?.offset ?? 0));
  return apiFetch<{ items: ErgonomicAssessmentOut[]; total: number }>("console", `/pgr/ergonomics?${qs.toString()}`);
}

export async function createErgonomicAssessment(payload: Record<string, any>) {
  return apiFetch<ErgonomicAssessmentOut>("console", "/pgr/ergonomics", { method: "POST", body: JSON.stringify(payload) });
}

export async function approveErgonomicAssessment(assessmentId: string, approval_notes?: string) {
  return apiFetch<ErgonomicAssessmentOut>("console", `/pgr/ergonomics/${assessmentId}/approve`, { method: "POST", body: JSON.stringify({ approval_notes }) });
}
